lib/spack/spack/build_environment.py

```diff
@@ -861,12 +866,16 @@ def setup_package(pkg, dirty, context='build'):
 
     implicit_rpaths = pkg.compiler.implicit_rpaths()
     if implicit_rpaths:
-        env.set('SPACK_COMPILER_IMPLICIT_RPATHS',
-                ':'.join(implicit_rpaths))
+        env_mods.set('SPACK_COMPILER_IMPLICIT_RPATHS',
+                     ':'.join(implicit_rpaths))
 
     # Make sure nothing's strange about the Spack environment.
-    validate(env, tty.warn)
+    validate(env_mods, tty.warn)
+
+    # Compose and apply the full environment modifications
+    env.extend(env_mods)
```

Comment:
I'm probably not familiar with this workflow - but if we are extending the clean env to include the mods ("not clean") what is the point of keeping the two separate? I guess I thought the point was to possibly save a cleaner version to file. So I think my questions are:

1. where does this returned environment (with both) go?
2. Is the one saved passed to `pkg.architecture.platform.setup_platform_environment(pkg, env_mods)` so it's ok?

Reply:
There is one "base" envmods (the clean env, or empty when --dirty), and one compiler/package modifications envmods. This has to be split, because we validate the package envmods for suspicious changes such as setting the compiler variable etc. These varables are also modified in the "base" envmods, so we shouldn't run validate on them.

In the end those two things are merged, and we log all the modifications together.